cpp/src/binaryop/compiled/operation.cuh

```diff
@@ -390,9 +390,11 @@ struct NullEquals {
     TypeLhs x, TypeRhs y, bool lhs_valid, bool rhs_valid, bool& output_valid) -> decltype(x == y)
   {
     output_valid = true;
-    if (!lhs_valid && !rhs_valid) return true;
-    if (lhs_valid && rhs_valid) return x == y;
-    return false;
+    if (!lhs_valid || !rhs_valid) return rhs_valid == lhs_valid;
+    if constexpr (cudf::is_floating_point<TypeLhs>() && cudf::is_floating_point<TypeRhs>()) {
+      if (isnan(x) && isnan(y)) return true;
+    }
+    return x == y;
```

Comment:
I think doing this unconditionally will break us. We usually want `NaN`s to compare not equal:

```python
In [11]: s = cudf.Series([1.0, 2.0, np.nan], nan_as_null=False)

In [12]: s == s
Out[12]: 
0     True
1     True
2    False
dtype: bool
```

Except when using `.equals()`, when we want NaNs to compare equal:

```
# s.equals(s.copy()) should return [True, True, True]
```

Reply:
Ok. I wasn't sure if coding this only in `NullsEquals` would you help here.